### ISSUE TITLE
Deletion race condition when fetching secret from the Kubernetes secrets provider cache

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#6340]
- Relates https://github.com/elastic/elastic-agent/issues/6187

The Kubernetes secrets provider caches secrets and updates them on a configurable schedule to avoid placing unnecessary load on the Kubernetes API.

The cache is updated in a goroutine: https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L88-L90

The cache is copied, the latest value for each non-expired secret is fetched, and then the copy is merged into the active map. https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L164-L199

The logic to get a value from the cache follows below.

https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L204-L225

1. A read lock on the cache is taken to check if the value exists in the cache.
2. If it isn't, the secret is fetched which involves holding a write lock again to update the cache.
3. A write lock on the cache is taken and the cache is updated with the value from step 2 and the last access time is set.

The race condition is that after 1 and 2 have completed, the `updateCache()` method that can delete secrets from the cache could have run and deleted the secret right before 3 below where the value is returned directly from the cache with no check for whether the secret still exists.

https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L218-L222

**Comment by cmacknz** [comment]
There is another potential problem here, where time since last access is evaluated twice:

Once here where we refresh the value if the TTL hasn't expired, but the problem is if the value changed and agent should update the component model with the new value, we don't update the last access.

https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L174-L182

The TTL is then evaluated again when merging the maps which can lead to deletion even though the value updated:

https://github.com/elastic/elastic-agent/blob/54932dc7b46cdd0d16791e07b8fc5dd596efb129/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L128-L130

This would only matter if the TTLDelete and the update time were very close but have an internal example of that.

**Comment by pkoutsovasilis** [comment]
> The design of this provider doesn't sit well with me. What we're trying to do is very simple: Keep an up-to-date local cache of a small number of Secrets. Intuitively, it'd be best to have an informer for these. However, informers (and watches in general) don't have great support for selecting resources based on field names. It's possible to create an informer which tracks a particular singular Secret by using a field selector, but if we want multiple Secrets, we need to have an informer for each of them, which puts some undue strain on the Kubernetes API Server.

yes AFAIK multiple FieldSelectors (aka selecting by name) is not a thing in k8s. Chained selectors is the closest thing to "multi-values" fieldselectors but they are chained with logical "and" and we need logical "or" here, thus unusable.

> If this was anything other than Secrets, I'd strongly consider using a single informer and just filtering within the application, but cluster operators get understandably touchy about applications wanting to access a lot of Secrets, so this doesn't sound like the best of ideas either.

As is now the agent RBAC requires just a "get" for just the secrets this provider has to "watch"; where with a single informer we would need "list", "watch" (and "get") for every secret out there. So just echoing the same message as the quote, not the best option

> Maybe we should at least use a battle-tested off-the-shelf component for this instead of implementing an expiration cache ourselves and opening ourselves up to concurrency bugs. https://pkg.go.dev/k8s.io/client-go/tools/cache#ExpirationCache looks like it would fit.

Utilising a battle-tested off-the-shelf component seems like the best way forward 🙂

**Comment by cmacknz** [comment]
> Utilising a battle-tested off-the-shelf component seems like the best way forward 🙂

+1 from me. TIL https://pkg.go.dev/k8s.io/client-go/tools/cache#ExpirationCache exists, thanks.
